### PR TITLE
Construct server capabilities using json-c

### PR DESCRIPTION
**Pull request by mnissler-rivos**
String formatting is hitting its limits: Adding another field is difficult given that we already branch on whether migration is enabled. This change constructs a JSON-C object instead so we can add what we need and serialize to a string afterwards.

### CODE REVIEW

File: lib/tran.c [hunk 1]
```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
+}
+
 static int
 send_version(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
              struct vfio_user_version *cversion)
 {
     struct vfio_user_version sversion = { 0 };
     struct iovec iovecs[2] = { { 0 } };
-    char server_caps[1024];
+    struct json_object* jo_caps = NULL;
```
Comment (jlevon) [inline lib/tran.c:351]: " *jo_caps" please

File: lib/tran.c [hunk 2]
```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
+    const char* server_caps = NULL;
```
Comment (jlevon) [inline lib/tran.c:352]: ditto

File: lib/tran.c [hunk 3]
```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
-                    "\"max_data_xfer_size\":%u,"
-                    "\"migration\":{"
-                        "\"pgsize\":%zu"
-                    "}"
-                "}"
-             "}", SERVER_MAX_FDS, SERVER_MAX_DATA_XFER_SIZE,
-                  migration_get_pgsize(vfu_ctx->migration));
+    int ret;
+
+    if ((jo_caps = create_caps_object(vfu_ctx)) == NULL ||
```
Comment (jlevon) [inline lib/tran.c:356]: nit, I would rather this function returns a strdup()ed string and the json handling is entirely within it.

File: lib/tran.c [hunk 4]
```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
+#endif
+
+    if (ret < 0) {
+        json_object_put(*val);
+    }
+    *val = NULL;
+    return ret;
+}
+
+static json_object* create_caps_object(vfu_ctx_t *vfu_ctx) {
```
Comment (jlevon) [inline lib/tran.c:294]: * placement

Comment (jlevon) [reply lib/tran.c:294]: function brace on new line please

File: lib/tran.c [hunk 5]
```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
+    struct json_object *jo_migration = NULL;
+    struct json_object *jo_caps = NULL;
+    struct json_object *jo_top = NULL;
+    struct json_object *jo_tmp = NULL;
+
+    if ((jo_caps = json_object_new_object()) == NULL) {
+        goto error;
+    }
+
+    if ((jo_tmp = json_object_new_int64(SERVER_MAX_FDS)) == NULL ||
```
Comment (jlevon) [inline lib/tran.c:304]: consider a json_add_int64() on top of json_add() since you do this several times?

File: lib/tran.c (hunk 1 above)
Comment (mnissler-rivos) [reply lib/tran.c:351]: fixing

File: lib/tran.c (hunk 2 above)
Comment (mnissler-rivos) [reply lib/tran.c:352]: fixing

File: lib/tran.c (hunk 4 above)
Comment (mnissler-rivos) [reply lib/tran.c:294]: fixing

File: lib/tran.c (hunk 3 above)
Comment (mnissler-rivos) [reply lib/tran.c:356]: That should be simpler indeed. Will do.

File: lib/tran.c (hunk 5 above)
Comment (mnissler-rivos) [reply lib/tran.c:304]: Good idea, will do.

**Review by jlevon — APPROVED** [review]

**Review by tmakatos — APPROVED** [review]
LGTM, just one nit

File: lib/tran.c [hunk 6]
```diff
@@ -265,36 +265,113 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
...
+json_add_uint64(struct json_object *jso, const char *key, uint64_t value)
+{
+    struct json_object *jo_tmp = NULL;
+
+    /*
+     * Note that newer versions of the library have a json_object_new_uint64
+     * function, but the int64 one is available also in older versions that we
+     * support, and our values don't require the full range anyways.
+     */
+    jo_tmp = json_object_new_int64(value);
```
[the remark this answers is not in the document]
Comment (mnissler-rivos) [reply lib/tran.c:304]: Yes, adding.

**Review by tmakatos — APPROVED** [review]

**Pull request merged**